gatsby/content/blog/2021/12/2021-12-10-sydent-typing-2.mdx

```diff
@@ -0,0 +1,387 @@
+- This repeats, until we eventually return an `IResponse`.
+
+(I'm guessing `x = await y` is really some kind of sugar around this mechanism: first `yield y`, then the machinery running your coroutine `c` will call `c.send(x)` to resume execution.)
+
+Why have I opted to use `Any` here, when we've seen (and will see) that this limits mypy's ability to run checks? The answer is that we yield two different types within the function. Firstly a [routing result](https://github.com/matrix-org/sydent/blob/e4b4dbbdf25255d19effceae7061ad9b3eefc374/sydent/http/matrixfederationagent.py#L152-L153):
+
+```python
+        routing: _RoutingResult
+        routing = yield defer.ensureDeferred(self._route_matrix_uri(parsed_uri))
+```
+
+and later, [the `IResponse`](https://github.com/matrix-org/sydent/blob/e4b4dbbdf25255d19effceae7061ad9b3eefc374/sydent/http/matrixfederationagent.py#L194-L195) we go on to return:
+
+```python
+        res: IResponse
+        res = yield agent.request(method, uri, headers, bodyProducer)
+```
+
+I could more correctly annotate the `YieldType` as `Deferred[Union[_RoutingResult, IResponse]]` so that the `SendType` was `Union[_RoutingResult, IResponse]`. But this would mean we end up having some kind of type check at each yield point: a `cast`, or a `type: ignore`, or a runtime `isinstance` check. It didn't feel like it was worth the boilerplate, especially since I could narrow e.g. `res: Any` to `res: IResponse` with minimal effort.
+
+This problem doesn't arise with using the more common `async/await` syntax:
+
+```python
+async def foo() -> int:
+    return 1
+
+async def bar() -> None:
+    x = await foo()
+    reveal_type(foo())
+    reveal_type(x)
+```
+
+```
+$ mypy example.py
+example.py:6: note: Revealed type is "typing.Coroutine[Any, Any, builtins.int]"
+example.py:7: note: Revealed type is "builtins.int*"
+```
+
+All in all, the handling of `inlineCallbacks` is a situation specific to working with (older?) twisted code. It's still nice to understand what's going on behind the scenes though!
+
+
+
+#### `zope.interface.Interface`
+
+Twisted makes use of `zope`'s `Interface` to define a number of abstract interface classes. Speaking personally, I've not seen it used outside twisted, and I think that means it's not supported by much of the typechecking tooling. For example, I've definitely seen PyCharm struggle to realise that it's okay to pass a `Response` to a function which expects an `IResponse`! Here's a more complicated example where `PyCharm` isn't happy with me widening the type `LoggingHostnameEndpoint` to `IStreamClientEndpoint`, even though [the latter implements the former](https://github.com/matrix-org/sydent/blob/92ff7a878a25696365b10cc49e32f5cba32c5960/sydent/http/matrixfederationagent.py#L379-L380).
+
+![Screenshot from pycharm showing a false positive warning](https://i.imgur.com/27rqjHm.png)
```

Comment:
I am not sure if this matters but the screenshot is not displaying on the preview, thought I would mention it!

Reply:
Thanks. Hopefully fixed with https://github.com/matrix-org/matrix.org/pull/1124/commits/07dfe911c31c35d6beefbd12769cb01b9c154f2e and https://github.com/matrix-org/matrix.org/pull/1124/commits/56dfb038652133178180d15bb44527ba709eeea1 and